File: 2400-number-of-ways-to-reach-a-position-after-exactly-k-steps/2400-number-of-ways-to-reach-a-position-after-exactly-k-steps.py
class Solution:
    def numberOfWays(self, startPos: int, endPos: int, k: int) -> int:
        dp = [[-1 for x in range(1001)] for y in range(1001)]
        mod = 1e9+7
        def go(k,dis):
            if(k<=dis):
                if(k==dis):
                    return 1
                return 0
            if(dp[k][dis]==-1):
                dp[k][dis] = (go(k-1,dis+1)+ go(k-1,abs(dis-1)))%mod
            return int(dp[k][dis])
        return int(go(k,abs(startPos-endPos)))

File: 2400-number-of-ways-to-reach-a-position-after-exactly-k-steps/2400-number-of-ways-to-reach-a-position-after-exactly-k-steps.cpp
class Solution {
public:
    unordered_map<int,long long>ump[1001];
    const long long mod = 1e9+7;
    int go(int a,int b,int c){
        if(ump[c].find(a)!=ump[c].end())
            return ump[c][a];
        if(c==0){
            if(a==b)return 1;
            return 0;
        }
        ump[c][a] = (go(a+1,b,c-1)+go(a-1,b,c-1))%mod;
        return ump[c][a];
    }
    int numberOfWays(int startPos, int endPos, int k) {
        // cout<<ump[0][0]<<endl;
        return (go(startPos+1,endPos,k-1) + go(startPos-1,endPos,k-1))%mod;
        // return 1;
    }
};